### PR TITLE
Math improvements

### PR DESCRIPTION
**Pull request by NogginBops**
### Purpose of this PR

Changes math types to use `HashCode` for hashcodes and VS autogenerated equality checks.

### Testing status

The math testing project passes all tests.

### CODE REVIEW

File: src/OpenTK.Mathematics/Data/Quaternion.cs [hunk 1]
```diff
@@ -799,58 +798,37 @@ public static Quaternion Slerp(Quaternion q1, Quaternion q2, float blend)
...
         /// </summary>
-        /// <param name="other">The other Quaternion to be used in the comparison.</param>
-        /// <returns>True if both instances are equal; false otherwise.</returns>
-        [Pure]
-        public bool Equals(Quaternion other)
+        /// <returns>A human-readable representation of the quaternion.</returns>
+        public override string ToString()
         {
-            return Xyz == other.Xyz && W == other.W;
+            return $"V: {Xyz}{MathHelper.ListSeparator} W: {W}";
```
Comment (NogginBops) [inline src/OpenTK.Mathematics/Data/Quaternion.cs:831]: Same thing here as with Color4, do we even want localized separators?

File: src/OpenTK.Mathematics/Data/Quaterniond.cs [hunk 2]
```diff
@@ -789,58 +788,37 @@ public static Quaterniond Slerp(Quaterniond q1, Quaterniond q2, double blend)
...
         /// </summary>
-        /// <param name="other">The other Quaterniond to be used in the comparison.</param>
-        /// <returns>True if both instances are equal; false otherwise.</returns>
-        [Pure]
-        public bool Equals(Quaterniond other)
+        /// <returns>A human-readable representation of the quaternion.</returns>
+        public override string ToString()
         {
-            return Xyz == other.Xyz && W == other.W;
+            return $"V: {Xyz}{MathHelper.ListSeparator} W: {W}";
```
Comment (NogginBops) [inline src/OpenTK.Mathematics/Data/Quaterniond.cs:821]: Same thing here as with Color4, do we even want localized separators?

**Review by NogginBops — COMMENTED** [review]
There are a few discussion points here relating to localization of `ToString`.

**Comment by varon** [comment]
This looks good, but we'll need benchmarks before we can merge to prove we haven't worsened performance. Hopefully there's some improvement too!

For localization, just do whatever vec3/vec4 were doing before. it's a minor issue for most users.

**Comment by jvbsl** [comment]
I don't really think we need perf tests for that. As far as I can tell the performance detriment should be minimal, and dependant on jitter even nonexistent, but the hashing is standardized and well implented. But for something like `Color` I would argue against using it and leaving the ToArgb part instead, as it is the optimal hashing we can probably get out of it without any collisions whatsoever.

**Comment by NogginBops** [comment]
`ToArgb` is not a perfrect hash fror `Color4` because `Color4` uses floats internally, while `ToArgb` casts down to `int` so for 8bpp colors it's a perfect hash but we can not assume 8bpp color because of things like HDR and similar. Having a proper hash function is going to give a better distribution and more expected characteristics for most cases.

**Comment by jvbsl** [comment]
my bad, I assumed Color4 uses bytes...then it is good and I'm fine with all the changes. About the seperators, I don't really know either

**Review by varon — APPROVED** [review]

**Comment by varon** [comment]
Thanks for this Noggin. Great work.

**Pull request merged**